README.md
## BitCaine5 Aeternity Multi CPU GPU Miner
BitCaine5 the mining engine meets æternity blockchain technology

have found a solution for multi multi cpu gpu mining!!!

NO            | NO                | NO                               | YES
------------- | ----------------- | -------------------------------- | -----------------
**NO DOCKER** | **NO EXTRA VM's** | **NO SOURCE CODE MANIPULATION**  | **only one node**
table: 1      |                   | **NO CONFIG (yaml)MODIFICATION** | 
## How does it work?
More next week, will talk to the devs on Monday

## Examples

COMPUTING      | AE NODES      | PRICE         | WORKS  | TESTED | OS     | LICENSING 
-------------- | ------------- | ------------- | ------ | ------ | ------ | -------
CPU  1 to  50  | ONE           | free          | YES    | YES    | LINUX  | GPL & PAID
GPU  1 to  50  | ONE           | free          | YES    | NO     | LINUX  | GPL & PAID
MIX  1 to  50  | TWO           | free          | YES    | NO     | LINUX  | GPL & PAID
CPU 50 to 100  | ONE           | need support  | YES    | NO     | LINUX  | GPL & PAID
GPU 50 to 100  | ONE           | need support  | YES    | NO     | LINUX  | GPL & PAID
MIX 50 to 100  | TWO           | need support  | YES    | NO     | LINUX  | GPL & PAID
+50 x CPU/GPU  | +1            | need support  | YES    | NO     | LINUX  | GPL & PAID
+50 x MIX      | +2            | need support  | YES    | NO     | LINUX  | GPL & PAID
table: 2

**as faster your server and as faster your network as less physical servers you need**

Mining APP     | PHYSICAL | VM
-------------- | -------- | -------
mean29-generic | YES      | YES
mean29-avx2    | YES      | DEPENDS
lean29-generic | YES      | YES
lean29-avx2    | YES      | DEPENDS
mean15-generic | YES      | YES
lean15-generic | YES      | YES
cuda29         | YES      | YES

**one node per mining app!!!**


### You have 1 to a million cPUs
* it will work !
* you need x nodes(see table 2) to manage them all, there a several solutions to run more 
than one EPOCH instance on one physical machine

### You have 1 to a million gPUs
* it will work !
* you need x nodes(see table 2) to manage them all, there a several solutions to run more 
than one EPOCH instance on one physical machine

 
---

### You will run a hole farm with a million CPUs and GPUs
* it will work !
* it depends how fast your servers and network are, test it you self and share you results

## How long does it take to setup
* less then a minute
* and +- one hour
* I will publish the scripts later

## Community Edition and paid Version
* full source code control

## Do I have to install any BitCain binary?
* **Yes! You have to build it always your self!!!**

## Why will I pay for the software if I can get for free?
* because I need your support to develop and you are not an idiot
* No support No software

donations are welcome: ak_XoongvC5xDqBCwdLr3ok1SzK3xEMFVwQ1sA7vj1guBd99HFjQ

## And yes the binary is not all
but before I received not one cent, I will not publish the rest of the how to's!!!



binaries/changelog.md
## Build Date and Time added into binaries
./cuda29_GTX_1070 -s

. . .

DEFAULTS
  ./cuda29_GTX_1070 -d 0 -E 0 -h "" -m 176 -n 0 -r 1 -U 4096 -u 256 -v 128 -w 512 -y 1024 -Z 1024 -z 1024
Build version  : cuda29_GTX_1070 by zwilla source_sha256: e99ea684f258c053f2bda842ddef176696ed0d87e22e9ee2e17469e76f94a231
Build date: 20181212-19:16:09

---

# Debug Mode reenabeled use -g at your extra_args: "-g" to see the output
* binaries updated

**Example without Debug mode:** (slower)
./cuda29_GTX_1070 -r 120 -h 324D6F373754576B3546684B772F493865493343416B71497531492B624A4B6155567A6543424B3278414D3D -n 18335313330172576564
Solution(fe741e54426bc345) 1406157 253890b 25caede 29f6a1d 2d981aa 31e596b 433e85c 6df4880 71e2c6c 80936e0 923bb91 94f0688 9838477 9d18eb8 9e11837 c0c997e cc7bf17 d235c3e d4ce55c d55c310 e05c39a e8f0a6c f4aa877 fc1364b 102028aa 12f4b651 13f38290 14a7724f 14aaa16f 15130716 17c36072 183de165 18b2d2ae 1ab0e11d 1af15642 1b00663b 1b149249 1e037d82 1e5e2580 1ec08efb 1ec584dd 1fea53bb
Solution(fe741e54426bc3a9) c15316 197817c 32bab3e 33dace8 4a0e47c 5ce2bdd 6c58999 94da3e2 992c1e1 9c79482 a3e6a5d b98452a bd1f19c c2c0be1 c85a493 cd107f3 e205020 ed3ccef 112d19b2 115386b7 119d35c4 1348e1ea 143cfc04 14487741 153c5b19 15407b7e 163bd23b 16a32c89 16ea3170 17d95218 19159e6b 1995a24d 1a77f7d9 1ace491e 1ae9de25 1bcee41c 1c5fb8b1 1cce89ce 1dcec3ac 1de2eea6 1eb82ffe 1efd3cf8
2 total solutions of 120 nonces

real	0m44.231s
user	0m24.681s
sys	  0m19.227s

**Example with Debug mode:** (faster)
Solution(fe741e54426bc3a9) c15316 197817c 32bab3e 33dace8 4a0e47c 5ce2bdd 6c58999 94da3e2 992c1e1 9c79482 a3e6a5d b98452a bd1f19c c2c0be1 c85a493 cd107f3 e205020 ed3ccef 112d19b2 115386b7 119d35c4 1348e1ea 143cfc04 14487741 153c5b19 15407b7e 163bd23b 16a32c89 16ea3170 17d95218 19159e6b 1995a24d 1a77f7d9 1ace491e 1ae9de25 1bcee41c 1c5fb8b1 1cce89ce 1dcec3ac 1de2eea6 1eb82ffe 1efd3cf8
Verified with cyclehash 368d2ab2220f12bc482db070e15eaca71493be3290bd66ed4750a1ec60b9e763
nonce 18335313330172576682 k0 k1 k2 k3 b4a41273c842792a 2118c0d57d6d4dae 766f9288dcf7945e ee8d443701cd5f1c
Seeding completed
   2-cycle found
  24-cycle found
  24-cycle found
 408-cycle found
 456-cycle found
findcycles edges 68233 time 21 ms total 347 ms
Time: 347 ms
nonce 18335313330172576683 k0 k1 k2 k3 a61f856e44931cac b34d3fe93b692c6e f318a26da07d49e8 444c48a3476ed103
Seeding completed
  14-cycle found
  12-cycle found
   6-cycle found
 536-cycle found
 420-cycle found
findcycles edges 64563 time 19 ms total 345 ms
Time: 345 ms
2 total solutions of 120 nonces

real	0m44.510s
user	0m24.635s
sys	0m19.556s

---

# Debug mode hard deactivated on last binaries
you will get just some lines like that:

<0.2010.13>@aec_pow_cuckoo:generate_int:203 Executing cmd: "./cuda29_GTX_1070 -h 324D6F373754576B3546684B772F493865493343416B71497531492B624A4B6155567A6543424B3278414D3D -n 18335313330172576564 -r 120 -g -d 1"
...
2018-12-12 14:27:15.201 [debug] <0.1778.13>@aec_pow_cuckoo:parse_generation_result:481 5 total solutions of 120 nonces

---

# will update new binaries next 20 mminutes
2018-12-12 14:27:14.482 [debug] <0.1774.13>@aec_pow_cuckoo:test_target:536 Hash of (<<50,193,194,157,209,166,90,131,129,57,17,212,40,148,210,142,51,65,92,57,135,75,227,54,116,253,165,127,199,18,189,211>>)
2018-12-12 14:27:14.483 [debug] <0.1774.13>@aec_pow_cuckoo:parse_generation_result:477 Failed to meet target (505185084)
2018-12-12 14:27:14.484 [debug] <0.1774.13>@aec_pow_cuckoo:test_target:536 Hash of (<<233,203,161,95,48,129,77,22,103,118,152,144,175,141,130,161,19,230,1,156,192,0,157,226,76,44,240,192,105,124,67,66>>)
2018-12-12 14:27:14.485 [debug] <0.1774.13>@aec_pow_cuckoo:parse_generation_result:477 Failed to meet target (505185084)
2018-12-12 14:27:14.486 [debug] <0.1774.13>@aec_pow_cuckoo:test_target:536 Hash of (<<14,43,169,90,164,250,19,182,8,98,121,46,3,247,136,238,98,140,255,135,15,163,227,247,205,109,160,169,186,227,96,68>>)
2018-12-12 14:27:14.486 [debug] <0.1774.13>@aec_pow_cuckoo:parse_generation_result:477 Failed to meet target (505185084)
2018-12-12 14:27:14.486 [debug] <0.1774.13>@aec_pow_cuckoo:parse_generation_result:481 3 total solutions of 120 nonces
2018-12-12 14:27:14.592 [debug] <0.1774.13>@aec_pow_cuckoo:generate:87 No cuckoo solution found
2018-12-12 14:27:14.636 [debug] <0.2010.13>@aec_pow_cuckoo:generate:81 Generating solution for data hash <<216,202,59,237,53,164,228,88,74,195,242,60,120,141,194,2,74,136,187,82,62,108,146,154,81,92,222,8,18,182,196,3>> and nonce 18335313330172576564 with target 505185084.
2018-12-12 14:27:14.639 [info] <0.2010.13>@aec_pow_cuckoo:generate_int:203 Executing cmd: "./cuda29_GTX_1070 -h 324D6F373754576B3546684B772F493865493343416B71497531492B624A4B6155567A6543424B3278414D3D -n 18335313330172576564 -r 120 -g -d 1"
2018-12-12 14:27:15.199 [debug] <0.1778.13>@aec_pow_cuckoo:test_target:536 Hash of (<<239,122,151,0,204,222,219,68,123,243,197,108,98,137,173,240,52,216,82,160,219,249,70,233,33,226,134,252,11,144,72,133>>)
2018-12-12 14:27:15.199 [debug] <0.1778.13>@aec_pow_cuckoo:parse_generation_result:477 Failed to meet target (505185084)
2018-12-12 14:27:15.199 [debug] <0.1778.13>@aec_pow_cuckoo:test_target:536 Hash of (<<196,96,151,217,70,156,239,57,184,34,223,16,168,108,35,243,217,252,41,186,237,169,110,235,140,21,125,108,146,212,34,107>>)
2018-12-12 14:27:15.199 [debug] <0.1778.13>@aec_pow_cuckoo:parse_generation_result:477 Failed to meet target (505185084)
2018-12-12 14:27:15.200 [debug] <0.1778.13>@aec_pow_cuckoo:test_target:536 Hash of (<<135,160,32,48,58,56,150,209,221,216,144,97,111,161,209,198,4,88,243,207,206,251,53,102,100,155,241,225,158,5,23,170>>)
2018-12-12 14:27:15.200 [debug] <0.1778.13>@aec_pow_cuckoo:parse_generation_result:477 Failed to meet target (505185084)
2018-12-12 14:27:15.200 [debug] <0.1778.13>@aec_pow_cuckoo:test_target:536 Hash of (<<100,160,217,214,170,197,14,58,56,55,127,171,122,72,216,119,180,35,169,209,48,185,109,155,77,57,176,106,170,98,234,82>>)

scripts/Ubuntu_Server/install_multi_gpu.sh
#!/bin/sh
##############################################################################
 #category "BitCaine5 Mining Engine for æternity blockchain technology"
 #package "BitCaine5 Scripts"
 #author Miguel Padilla <miguel.padilla@zwilla.de>
 #copyright (c) 2012 - 2018 Miguel Padilla
 #link "https://BitCaine5.com"
 #github "https://github.com/Zwilla/BitCaine5_aeternity_miner"
 #twitter "https://twitter.com/mytokenwallet"
 #license: see LICENSE - > dual licensing
##############################################################################
 # This program is distributed in the hope that it will be useful,
 # but WITHOUT ANY WARRANTY; without even the implied warranty of
 # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 # License for more details.
 # FILE: install_multi_gpu.sh
 
 exit;
 # Do not execute this file, use it to get a valid build on your system, and yes there are many other things you need
  
# check if your cuda driver is intalled correct:

/usr/bin/nvidia-smi
 
# you will get this
# Wed Dec 12 01:00:51 2018       
# +-----------------------------------------------------------------------------+
# | NVIDIA-SMI 410.79       Driver Version: 410.79       CUDA Version: 10.0     |
# |-------------------------------+----------------------+----------------------+
# | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
# | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
# |===============================+======================+======================|
# |   0  GeForce GTX 107...  Off  | 00000000:11:00.0 Off |                  N/A |
# |  0%   43C    P0    36W / 180W |      0MiB /  8119MiB |      0%      Default |
# +-------------------------------+----------------------+----------------------+
# |   1  GeForce GTX 107...  Off  | 00000000:90:00.0 Off |                  N/A |
# |  0%   42C    P0    34W / 180W |      0MiB /  8119MiB |      0%      Default |
# +-------------------------------+----------------------+----------------------+
                                                                               
# +-----------------------------------------------------------------------------+
# | Processes:                                                       GPU Memory |
# |  GPU       PID   Type   Process name                             Usage      |
# |=============================================================================|
# |    0     35749      C   ./cuda29_GTX_1070                             13MiB |
# |    1     35752      C   ./cuda29_GTX_1070                             13MiB |
# +-----------------------------------------------------------------------------+
 

 
 # if your node will not stop or you are unable to stop it use this command
 # killall -9 run_erl
 # normal way to stop is
 # $HOME/multi_gpu_final/_build/prod/rel/epoch/bin/epoch stop
 
cd ~
git clone https://github.com/aeternity/epoch.git multi_gpu_final && cd multi_gpu_final
git checkout multi_gpu_final
make prod-build

#now we make the driver
cd $HOME/multi_gpu/apps/aecuckoo
make
#simple test
time $HOME/multi_gpu/apps/aecuckoo/priv/bin/cuda29 -r 50000

# now we need the optimized cuda version
# example: GTX_1070 and GTX_1070i are the same so use GTX_1070
YOUR_CARD_Series=GTX_1070
cd $HOME/multi_gpu_final/_build/prod/rel/epoch/lib/aecuckoo-0.1.0/priv/bin
wget https://raw.githubusercontent.com/Zwilla/BitCaine5_aeternity_miner/master/binaries/cuda29_$YOUR_CARD_Series
chmod a+x $HOME/multi_gpu_final/_build/prod/rel/epoch/lib/aecuckoo-0.1.0/priv/bin/cuda29_$YOUR_CARD_Series

#validate it
$HOME/multi_gpu_final/_build/prod/rel/epoch/lib/aecuckoo-0.1.0/priv/bin/cuda29_$YOUR_CARD_Series -s
# SYNOPSIS
# cuda29_GTX_1070 
# [-d device] 
# [-E 0-2] 
# [-h hexheader] 
# [-m trims] 
# [-n nonce] 
# [-r range] 
# [-U seedAblocks] 
# [-u seedAthreads] 
# [-v seedBthreads] 
# [-w Trimthreads] 
# [-y Tailthreads] 
# [-Z recoverblocks] 
# [-z recoverthreads] 
# [-g debug] 
# [-c cpu none blocking]

# DEFAULTS
#  cuda29_GTX_1070 -d 0 -E 0 -h "" -m 176 -n 0 -r 1 -U 4096 -u 256 -v 128 -w 512 -y 1024 -Z 1024 -z 1024
# Build version  : cuda29_GTX_1070 by zwilla source_sha256: 504e57f73878e274f3875fbe56fad32e0f7dd3254a75059272d6b4cd0b030acf
# Build date: 20181212

#test it
time $HOME/multi_gpu_final/_build/prod/rel/epoch/lib/aecuckoo-0.1.0/priv/bin/cuda29_$YOUR_CARD_Series -g -r 600 -c
time $HOME/multi_gpu_final/_build/prod/rel/epoch/lib/aecuckoo-0.1.0/priv/bin/cuda29_$YOUR_CARD_Series -g -r 600

# now you will get some results if you have a shorter time on -c then set your extra_args to "-c"

# copy your keys to this folder

ulimit -n 24576 && $HOME/multi_gpu_final/_build/prod/rel/epoch/bin/epoch start
